### PR TITLE
Bluetooth: Controller: Fix pdu_free_sem_give assertion under ZLI use

### LINKED ISSUE #46558: Bluetooth: Controller: Crash on bt_le_adv_start() when using CONFIG_BT_CTLR_ADVANCED_FEATURES
We need to enable `CONFIG_BT_CTLR_ADVANCED_FEATURES` because we need to adjust `CONFIG_BT_CTLR_ULL_HIGH_PRIO`, see https://github.com/zephyrproject-rtos/zephyr/discussions/41401#discussioncomment-1858153

In Zephyr `3.1.0` this causes a crash in `bt_le_adv_start()`:

```[00:00:00.411,926] <inf> bt: Starting Advertisement...

ASSERTION FAIL [!retval] @ WEST_TOPDIR/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:802


[00:00:00.422,210] <err> os: ***** HARD FAULT *****
[00:00:00.422,210] <err> os:   Fault escalation (see below)
[00:00:00.422,241] <err> os: ARCH_EXCEPT with reason 3

[00:00:00.422,241] <err> os: r0/a1:  0x00000003  r1/a2:  0x00000001  r2/a3:  0x00000000
[00:00:00.422,241] <err> os: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x0003ead7
[00:00:00.422,271] <err> os:  xpsr:  0x61000028
[00:00:00.422,271] <err> os: Faulting instruction address (r15/pc): 0x0003eae2
[00:00:00.422,302] <err> os: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
[00:00:00.422,332] <err> os: Fault during interrupt handling

[00:00:00.422,363] <err> os: Current thread: 0x20015f90 (unknown)
[00:00:01.841,278] <err> os: Halting system
```

Note: In Zephyr `2.7.2` same code/config worked fine. I think also on Zephyr `3.0.0` it was still ok (tbc.).

@cvinayak FYI

### PR DESCRIPTION
**Pull request by cvinayak**
Fix assertion due to multiple mayfly_enqueue calls used
under ZLI when pdu_free_sem_give is invoked from the LLL.

Fixes #46558.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

**Review by carlescufi — DISMISSED** [review]

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c [hunk 1]
```diff
@@ -817,6 +817,33 @@ static void pdu_free_sem_give(void)
...
 
+static void pdu_free_sem_take_all_no_wait(void)
+{
+	int err;
+
+	/* Flush all given semaphore count due to multiple release from LLL
+	 * context.
+	 */
+	do {
+		err = k_sem_take(&sem_pdu_free, K_NO_WAIT);
```
Comment (carlescufi) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:828]: Why not `k_sem_reset()` ?

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:828]: Was not aware of this! ... Let @caco3 verify the current fix, I will replace it with `k_sem_reset()` and verify again.

**Comment by caco3** [comment]
Thanks @cvinayak for the quick fix!
Your fix indeed fixes my issue!

**Comment by cvinayak** [comment]
> Thanks @cvinayak for the quick fix! Your fix indeed fixes my issue!

Thank you, @caco3. Please test once more with the change to use k_sem_reset(), we can then merge it upstream as soon as CI passes and have your confirmation.

**Comment by caco3** [comment]
I can confirm that `4884840` also works.

**Review by carlescufi — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**